bson/src/main/org/bson/codecs/DecoderContext.java

```diff
@@ -29,14 +32,40 @@
 public final class DecoderContext {
     private static final DecoderContext DEFAULT_CONTEXT = DecoderContext.builder().build();
     private final boolean checkedDiscriminator;
+    private final Map<String, Object> attachments;
 
+    /**
+     * Attaches an item to this context with the given name.  Returns a new modified context.
+     *
+     * @param key the key for the attachment
+     * @param object the item to attach
+     * @return the new DecoderContext
+     */
+    public DecoderContext attach(String key, Object object) {
+        return builder()
+                .attachments(attachments)
+                .checkedDiscriminator(checkedDiscriminator)
```

Comment:
these two lines, e.g., could be in a private "copy constructor" on `Builder` that handles copying the state.